autode/hessians.py

```diff
@@ -312,3 +317,200 @@ def frequencies_proj(self) -> List[Frequency]:
+    Arguments:
+        species: Species to calculate the numerical Hesssian for
+
+        *args: Arguments passed to the calculator
+
+        **kwargs: Keyword arguments passed to the calculator
+
+    Raises:
+        (ValueError): For unsupported input
+    """
+
+    calculator = _NumericalHessianCalculator(species, *args, **kwargs)
+    calculator.calculate()
+
+    species.hessian = calculator.hessian
+
+    return None
+
+
+class _NumericalHessianCalculator:
+
+    def __init__(self,
+                 species:   'autode.species.Species',
+                 method:    'autode.wrappers.base.Method',
+                 keywords:  'autode.wrappers.keywords.GradientKeywords',
+                 do_c_diff:  bool,
+                 num_delta: 'autode.values.Distance'
+                 ):
+
+        self._species = species
+        self._method = method
+        self._keywords = self._validated(keywords)
+
+        self._do_c_diff = do_c_diff
+        self._num_delta = num_delta
+
+        self._hessian = Hessian(np.zeros(shape=self._hessian_shape),
+                                units='Ha Å^-2')
+        self._calculated_rows = []
+
+        self._n_cores = max(Config.n_cores // self._n_rows, 1)
+
+    @work_in('numerical_hessian')
+    def calculate(self) -> None:
+        """Calculate the Hessian"""
+        logger.info(f'Calculating a numerical Hessian'
+                    f'{"with" if self._do_c_diff else "without"} central '
+                    f'differences using {self._n_cores} per process')
+
+        if not self._do_c_diff:
+            logger.info('Calculating gradient at current point')
+            self._init_gradient = self._gradient(species=self._species)
+
+        # Although n_rows may be < n_cores there will not be > n_rows processes
+        with Pool(processes=Config.n_cores) as pool:
+
+            func_name = '_cdiff_row' if self._do_c_diff else '_diff_row'
+
+            res = [pool.apply_async(hashable(func_name, self),
+                                    args=(i, k))
+                   for (i, k) in self._idxs_to_calculate()]
+
+            for row_idx, row in enumerate(res):
+                self._hessian[row_idx, :] = row.get(timeout=None)
+
+        return None
+
+    @property
+    def hessian(self) -> Hessian:
+        return self._hessian
+
+    @property
+    def _hessian_shape(self) -> Tuple[int, int]:
+        """Shape of the Hessian matrix for the species"""
+        return 3 * self._species.n_atoms, 3 * self._species.n_atoms
+
+    @property
+    def _n_rows(self) -> int:
+        """Number of rows in the Hessian"""
+        return 3 * self._species.n_atoms
+
+    def _new_species(self, atom_idx: int, component: int, direction: str):
+        """
+        New species with an applied shift to an atom. For example, water_0x+
+        for a water molecule where the 0th atom has been shifted in the
+        positive x direction
+        """
+        assert direction in ('+', '-')    # Positive or negative
```

Comment:
you know my feelings about strings with meanings :P

Reply:
I know, I know